Retry command invocation with argfile

weihanglo opened the pull request:
### What does this PR try to resolve?

Fixes #10381

The goal is to invoke `rustc` and `rustdoc` with `@path` arg file as a fallback.

Since the `-Zcheck-cfg-features`[^1] has already been implemented, in the foreseeable future cargo will probably hit the “**command line too big**” more often on Windows. `rustdoc` and `rustc` both support `@path` argfile [^2][^3], so we can leverage the feature for the fallback logic.

The idea behind this PR is that we put the **retry logic** in `ProcessBuilder::exec*` functions and reuse them as much as possible.

### How should we test and review this PR?

Review it commit by commit is recommended.

Argfile fallback is hard to tested with integration tests, so I added some unit tests for `cargo_util::ProcessBuilder` and `cargo::ops::fix::FixArgs`.

If you do want to test the support of argfile manually, a new environment variable `__CARGO_TEST_FORCE_ARGFILE` is added for the sake of forcing cargo to use argfile for rustc invocations. For example, `__CARGO_TEST_FORCE_ARGFILE cargo test --test testsuite -- fix::` is usually a good start to test this feature.

[^1]: https://doc.rust-lang.org/beta/cargo/reference/unstable.html?highlight=check#check-cfg-features
[^2]: https://doc.rust-lang.org/rustc/command-line-arguments.html#path-load-command-line-flags-from-a-path
[^3]: https://doc.rust-lang.org/rustdoc/command-line-arguments.html#path-load-command-line-flags-from-a-path

<!-- homu-ignore:start -->
### Additional information

Should we escape line feed when writing into argfile and unescape while reading? 82781929b8fffd65f363aab1549e1b8f6773920f implements that but I am not sure whether it is necessary.
<!-- homu-ignore:end -->

rust-highfive commented:
r? @ehuss

(rust-highfive has picked a reviewer for you, use r? to override)

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
+            .tempfile()?;
+
+        let path = tmp.path().display();
+        let mut cmd = self.build_command_without_args();
+        cmd.arg(format!("@{path}"));
```
Observation: This won't break wrappers that don't understand argfiles because they already weren't spawning.

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
+
+        match cmd.spawn() {
+            Err(ref e) if self.should_retry_with_argfile(e) => {
+                let (mut cmd, _argfile) = self.build_command_with_argfile()?;
+                apply(&mut cmd);
+                cmd.spawn()
```
Aren't we closing / removing the tempfile before the process has completed, racing with the application that is accessing it?

And should we explicitly close the tempfile, identifying any errors from the clean up?

weihanglo replied:
Fixed with 9155c0062bb6fcde2d8d727705841327dbe56e3f. It also removes the awkward `build_and_spawn` function as a side effect.

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -398,23 +496,35 @@ impl ProcessBuilder {
+
+        let mut error = command.exec();
+        if process_builder.should_retry_with_argfile(&error) {
+            let (mut command, _argfile) = process_builder.build_command_with_argfile()?;
+            error = command.exec()
```
While this doesn't have the race condition between the process accessing the tempfile and us closing it, should we check for an error with close?

epage commented on src/cargo/util/rustc.rs:
```diff
@@ -93,18 +93,24 @@ impl Rustc {
+        let mut cmd = ProcessBuilder::new(self.path.as_path()).wrapped(self.wrapper.as_ref());
+        cmd.retry_with_argfile(true);
+        cmd
     }
 
     /// Gets a process builder set up to use the found rustc version, with a wrapper if `Some`.
     pub fn workspace_process(&self) -> ProcessBuilder {
-        ProcessBuilder::new(self.path.as_path())
+        let mut cmd = ProcessBuilder::new(self.path.as_path())
             .wrapped(self.workspace_wrapper.as_ref())
-            .wrapped(self.wrapper.as_ref())
+            .wrapped(self.wrapper.as_ref());
+        cmd.retry_with_argfile(true);
+        cmd
     }
 
     pub fn process_no_wrapper(&self) -> ProcessBuilder {
-        ProcessBuilder::new(&self.path)
+        let mut cmd = ProcessBuilder::new(&self.path);
+        cmd.retry_with_argfile(true);
+        cmd
     }
```
How comes `retry_with_argfile` isn't chained

weihanglo replied:
I followed the existing pattern from both `ProcessBuilder` and `std::process::Command`. It is worth making it more chainable, but a dedicated PR should better IMO.

epage commented on src/cargo/ops/fix.rs:
```diff
@@ -774,35 +777,65 @@ struct FixArgs {
+            for arg in argv {
+                handle_arg(arg)?;
+            }
         }
```
Side note: While I'm assuming this is doing things like in other places, there is now the [argfile crate](https://docs.rs/argfile/latest/argfile/)

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -550,14 +547,87 @@ mod imp {
+                match c {
+                    '\n' => s.push_str("\\n"),
+                    '\r' => s.push_str("\\r"),
+                    '\\' => s.push_str("\\\\"),
+                    c => s.push(c),
+                }
+            }
+            Cow::Owned(s)
+        } else {
+            Cow::Borrowed(&self.arg)
+        }
+    }
+
+    /// Unescapes the argument.
+    ///
+    /// - `\\n` -> `\n`
+    /// - `\\r` -> `\r`
+    /// - `\\\\` -> `\\`
+    /// - `\\<any>` -> `<any>`
+    pub fn unescape(&self) -> Cow<'_, str> {
+        if self.arg.contains('\\') {
+            let mut s = String::with_capacity(self.arg.len() - 1);
+            let mut chars = self.arg.chars();
+            while let Some(c) = chars.next() {
+                match c {
+                    '\\' => match chars.next() {
+                        Some('n') => s.push('\n'),
+                        Some('r') => s.push('\r'),
+                        Some(c) => s.push(c),
+                        None => {}
+                    },
+                    c => s.push(c),
+                }
+            }
+            Cow::Owned(s)
+        } else {
+            Cow::Borrowed(&self.arg)
+        }
+    }
+}
```
I'm always paranoid of doing escaping wrong.  Is there an existing escaping mechanism in the stdlib that would be useful?

weihanglo replied:
Yes there is a [`std::ascii::escape_default`](https://doc.rust-lang.org/stable/std/ascii/fn.escape_default.html) but we need an unescape counterpart and it seems not existing. Will remove this whole escaping part as your advice. See https://github.com/rust-lang/cargo/pull/10546#discussion_r846788690

epage commented on src/cargo/ops/fix.rs:
```diff
@@ -822,17 +822,20 @@ impl FixArgs {
             rustc = iter
                 .next()
-                .map(PathBuf::from)
+                .map(|a| PathBuf::from(a.unescape().as_ref()))
```
Even if we don't use argfile, it seems like abstracting this into a function that handles the escaping would be helpful

weihanglo replied:
Sorry I don't understand what to abstract. Can you say it a bit clear?

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
+                    )
+                })?;
+            // TODO: Shall we escape line feed?
+            buf.push_str(arg);
```
Looks like a later commit addressed this

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -411,21 +412,17 @@ impl ProcessBuilder {
+                    "argument contains invalid UTF-8 characters",
+                )
+            })?;
+            writeln!(buf, "{}", ArgEscape::from(arg))?;
```
I guess the other question to escaping is what do rustc and similar tools support?

weihanglo replied:
This is [how rustc generates argfiles for ld](https://github.com/rust-lang/rust/blob/8c1fb2eb23964b56bd279b12aa925935d8559870/compiler/rustc_codegen_ssa/src/back/link.rs#L1442-L1480). As I read it so far, they just follow the specification. rustc uses only newlines to determine so I guess escaping `\r` `\n` and `\\` might be sufficient.

However, I found that [rustc_driver just naively reads line by line from argfile](https://github.com/rust-lang/rust/blob/1.60.0/compiler/rustc_driver/src/args.rs#L7-L18), they won't do any unescaping. As long as rustc does nothing special and our primary use case is serving rustc, I decide to **remove the escape part and error out** if someone passes line feeds in, as you recommended earlier. 

BTW there is [a test case](https://github.com/rust-lang/cargo/blob/5f400f098ba6bd76eb57c11bbe7c39a443146afe/tests/testsuite/rustdocflags.rs#L101-L123) validating that behaviour, which unfortunately accept whitespaces in `RUSTDOCFLAGS`.

weihanglo commented:
Addressed issues in review comments. Ready to re-review :)

- Drop escape/unescape (commit 82781929b8fffd65f363aab1549e1b8f6773920f) because rustc doesn't understand how to unescape escaped characters.
- Add `__CARGO_TEST_FORCE_ARGFILE` env var to force using argfile for rustc invocation in debug mode. 5e8827f26e599f249da01db8cf8ac6ae195f3bb8
- Add a CI test to check argfile operability by force rustc to use `@path` argfile. Only test `cargo fix` because it is the only place reading from argfile. 275b0c68d4157045bc17341ff31422f1248c1dc0

weihanglo commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
+            .prefix("cargo-argfile.")
+            .tempfile()?;
+
+        let path = tmp.path().display();
```
Cargo conventionally constructs argument for rustc with OsString, so I feel like we should  and it has been fixed with 3f749f615b6bd5afcb076fcb623f48c9f1c5701e. (Though rustc [rejects to handle non Unicode arguments 🥲](https://github.com/rust-lang/rust/blob/1.60.0/compiler/rustc_driver/src/lib.rs#L1319-L1329))

epage commented on crates/cargo-util/src/process_builder.rs:
```diff
@@ -363,46 +510,80 @@ impl ProcessBuilder {
 
-        let args = once(self.program).chain(self.args.into_iter()).collect();
+/// Forces the command to use `@path` argfile.
+///
+/// You should set `__CARGO_TEST_FORCE_ARGFILE` to enable this.
+fn debug_force_argfile(retry_enabled: bool) -> bool {
+    cfg!(debug_assertions) && env::var("__CARGO_TEST_FORCE_ARGFILE").is_ok() && retry_enabled
+}
 
```
Observation: I could see in the future providing a flag in release mode as a way for people to workaround or debug problems.

epage reviewed (approved):
Looks good!

epage commented:
@bors r+

bors commented:
:pushpin: Commit 275b0c68d4157045bc17341ff31422f1248c1dc0 has been approved by `epage`

<!-- @bors r=epage 275b0c68d4157045bc17341ff31422f1248c1dc0 -->
<!-- homu: {"type":"Approved","sha":"275b0c68d4157045bc17341ff31422f1248c1dc0","approver":"epage"} -->

bors commented:
:hourglass: Testing commit 275b0c68d4157045bc17341ff31422f1248c1dc0 with merge 1073915930bf36f39472b0d223c125a3b51a9080...
<!-- homu: {"type":"BuildStarted","head_sha":"275b0c68d4157045bc17341ff31422f1248c1dc0","merge_sha":"1073915930bf36f39472b0d223c125a3b51a9080"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/5975968228?check_suite_focus=true)
Approved by: epage
Pushing 1073915930bf36f39472b0d223c125a3b51a9080 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/5975968228?check_suite_focus=true"},"merge_sha":"1073915930bf36f39472b0d223c125a3b51a9080"} -->
